Modules/_sqlite/connection.c

```diff
@@ -1761,29 +1754,29 @@ pysqlite_connection_create_collation_impl(pysqlite_Connection *self,
     if (!uppercase_name_str)
         goto finally;
 
-    if (callable != Py_None && !PyCallable_Check(callable)) {
-        PyErr_SetString(PyExc_TypeError, "parameter must be callable");
-        goto finally;
+    int flags = SQLITE_UTF8;
+    if (callable == Py_None) {
+        rc = sqlite3_create_collation_v2(self->db, uppercase_name_str, flags,
+                                         NULL, NULL, NULL);
     }
-
-    if (callable != Py_None) {
-        if (PyDict_SetItem(self->collations, uppercase_name, callable) == -1)
-            goto finally;
-    } else {
-        if (PyDict_DelItem(self->collations, uppercase_name) == -1)
+    else {
+        if (!PyCallable_Check(callable)) {
+            PyErr_SetString(PyExc_TypeError, "parameter must be callable");
             goto finally;
+        }
+        rc = sqlite3_create_collation_v2(self->db, uppercase_name_str, flags,
+                                         Py_NewRef(callable),
+                                         &pysqlite_collation_callback,
+                                         &_destructor);
     }
 
-    rc = sqlite3_create_collation(self->db,
-                                  uppercase_name_str,
-                                  SQLITE_UTF8,
-                                  (callable != Py_None) ? callable : NULL,
-                                  (callable != Py_None) ? pysqlite_collation_callback : NULL);
     if (rc != SQLITE_OK) {
+        /* Unlike other sqlite3_* functions, the destructor callback is _not_
+         * called if sqlite3_create_collation_v2() fails, so we have to free
+         * the context before returning.
+         */
         if (callable != Py_None) {
-            if (PyDict_DelItem(self->collations, uppercase_name) < 0) {
-                PyErr_Clear();
-            }
+            _destructor(callable);
```

Comment:
_descructor acquires the GIL and does a DECREF. Here we hold the GIL, so just a DECREF will do.

Reply:
Ah, thanks! Previously it was just a decref, but Pablo recently added the GIL wrapper. I had forgotten.

Reply:
> I see that the beginning of the function does some busywork converting `name` to uppercase and checking for allowed characters. Is that limitation there only to make that sure comparison of keys in the `collations` dict match `sqlite3_strnicmp`? If so, `PyUnicode_AsUTF8(name)` could be used instead.

Yes, I noticed that part as well. I haven't investigated why this was added in the first place. I'll try with `PyUnicode_AsUTF8()`. Thanks!

> As for the `finally`→`error` rename, I'd say `finally` is a better name when it's also executed in the non-error case...

True.